pkg/cloudprovider/providers/vsphere/vsphere.go

```diff
@@ -1374,3 +1656,107 @@ func makeDirectoryInDatastore(c *govmomi.Client, dc *object.Datacenter, path str
 
 	return err
 }
+
+// Delete the folder identify by path in the datacenter.
+func deleteDatastoreDirectory(ctx context.Context, c *govmomi.Client, dc *object.Datacenter, path string) error {
+	fileManager := object.NewFileManager(c.Client)
+
+	task, err := fileManager.DeleteDatastoreFile(ctx, path, dc)
+	if err != nil {
+		return err
+	}
+
+	return task.Wait(ctx)
+}
+
+// Gets the folder path from the VM Disk path.
+func getFolderFromVMDiskPath(vmDiskPath string) string {
+	i := strings.Index(vmDiskPath, "/")
+	folderPath := ""
+	if i > -1 {
+		folderPath = vmDiskPath[:i+1]
+	}
+	return folderPath
+}
+
+// Delete the VM.
+// An attempt is made to delete the VM for DeleteVMAttempts times before it backs off.
```

Comment:
I was not expecting a solution like this since the deleteVM could also fail 5 times (probably because of network issues). Can we not start a background cleanup task that can discover such VMs and cleanup? Btw, if we follow the other approach I proposed above (create a "kubevols" dummy VM once per datastore), then this problem is non-existent.